backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/impl/LuceneNestedDocumentsSort.java

```diff
@@ -32,4 +47,50 @@ public void add(LuceneNestedDocumentFieldContribution contribution) {
 	public boolean isEmpty() {
 		return contributions.isEmpty();
 	}
+
+	public void processNestedPaths(Query luceneQuery, IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		for ( Map.Entry<String, Set<LuceneNestedDocumentFieldContribution>> fieldEntry : contributions.entrySet() ) {
+			String nestedDocumentPath = fieldEntry.getKey();
+			nestedDocumentMaps.put( nestedDocumentPath, getFetchNestedDocumentMap( nestedDocumentPath, luceneQuery, indexSearcher, scoreDocs ) );
```

Comment:
So I guess you intend to improve on this, since one query + one query per sort doesn't look too good. Two I can understand, but 1 + N is a bit extreme.

Can't the `LuceneChildrenCollector` be adapted to produce multiple maps? Whenever it inspects a document, it could inspect its "nested document path" and decide to which map it should go based on that.

Reply:
Currently we do one query for each nested document paths that is present on some sort field. The same result is reused among all fields on the same nested object. Actually I don't know if there is a way to improve that. Since different nested document paths can have different nested documents. If there was, I would happy to improve that.

Reply:
From what I've got Elasticsearch does one ParentChildrenBlockJoinQuery for each parent document. That should be definitely worse.

Reply:
I'm not sure we're talking about the same thing. As far as I can tell, `org.elasticsearch.search.sort.SortBuilder#resolveNestedQuery` is used to produce the query that will return all nested documents, 

You're right that they don't seem to attempt to run one query for everything, though: apparently they just run `org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested#innerDocs` once for each sort, which is similar to what you're doing.

I suspect their use of a DocIdSet is much more efficient than the Map you're using, though. In particular they seem to retrieve nested documents... on the fly!

* `org.elasticsearch.search.sort.SortBuilder#resolveNested(org.elasticsearch.index.query.QueryShardContext, org.elasticsearch.search.sort.NestedSortBuilder)` creates, for a given query, a given sort and a given list of nested paths, a `Nested` utility object. I think it's similar to your `LuceneNestedDocumentFieldContribution`, except it has knowledge of the query.
* `org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested#innerDocs` returns, for a given `LeafReaderContext` (~a segment of the index), a `DocIdSetIterator` that will return all relevant nested documents.
* `org.apache.lucene.search.FieldComparator.TermValComparator#getBinaryDocValues` (and similar classes for other types of DocValues) wraps the DocValues to be used for sorts with a wrapper that uses `MultiValueMode#select`. It assumes these wrapped DocValues will transparently redirect to nested documents when the value of the parent document is requested.
* `org.elasticsearch.search.MultiValueMode#select(org.elasticsearch.index.fielddata.SortedBinaryDocValues, org.apache.lucene.util.BytesRef, org.apache.lucene.util.BitSet, org.apache.lucene.search.DocIdSetIterator, int, int)` (and similar methods for other types of DocValues)  takes advantage of the nested documents "docIdSet" to make sure that whenever someone requests the docvalues to advance to parent document, the doc values will instead advance to the relevant nested document if necessary.